Allow overriding the residue item in PlayerItemConsumeEvent

Wraith0x10 opened the pull request:
This PR allows you to override the residue item given in PlayerItemConsumeEvent. 
Such as removing the empty bottle residue when drinking a potion. 

### API changes
- Added `PlayerItemConsumeEvent->getResidue()`
- Added `PlayerItemConsumeEvent->setResidue()`

## Tests
This test removes the empty water bottle when consuming a potion.
```php
    public function onItemConsume(PlayerItemConsumeEvent $event): void {
        if($event->getItem() instanceof Potion) {
            $event->setResidue(VanillaItems::AIR());
        }
    }
```

Commit:
Allow overriding the residue item in PlayerItemConsumeEvent

GroXzaLYs commented on src/event/player/PlayerItemConsumeEvent.php:
```diff
@@ -44,4 +46,13 @@ public function __construct(
+		return $this->residue === null ? null : clone $this->residue;
+	}
+
+	/** Allows overriding the residue item (or null to remove residue) */
```
Delete this comment

Commit:
Satisfy PHPStan

dktapps commented on src/player/Player.php:
```diff
@@ -1740,7 +1741,8 @@ public function consumeHeldItem() : bool{
 			$slot->pop();
-			$this->returnItemsFromAction($oldItem, $slot, [$slot->getResidue()]);
+			$residue = $ev->getResidue();
+			$this->returnItemsFromAction($oldItem, $slot, $residue !== null ? [$residue] : []);
```
Maybe worth putting an array of items in the event, then plugins can set [] for no items, or potentially return multiple items if that's needed in the future

Wraith0x10 replied:
Done, it now accepts an array of items. 
I've tested it in game:

```php
    public function onItemConsume(PlayerItemConsumeEvent $event): void {
        $item = $event->getItem();
        if($item instanceof Potion){
            $event->setResidue([]);
        } elseif ($item instanceof GoldenApple){
            $event->setResidue([VanillaItems::POTATO(), VanillaItems::BOWL()]);
        }
    }
```

Commit:
Switch residue items to an array

dktapps commented on src/event/player/PlayerItemConsumeEvent.php:
```diff
@@ -47,12 +49,20 @@ public function getItem() : Item{
+	 */
+	public function setResidue(array $items) : void{
+		Utils::validateArrayValueType($items, function(Item $_) : void{});
+		$this->residue = $items;
```
These need to be cloned

dktapps commented on src/event/player/PlayerItemConsumeEvent.php:
```diff
@@ -47,12 +49,20 @@ public function getItem() : Item{
+	 * @return Item[]
+	 */
+	public function getResidue() : array{
+		return $this->residue;
```
Same here

Wraith0x10 replied:
Done

Commit:
Clone items

kostamax27 commented on src/event/player/PlayerItemConsumeEvent.php:
```diff
@@ -44,4 +48,32 @@ public function __construct(
+	/**
+	 * @return Item[]
+	 */
+	public function getResidue() : array{
```
`Utils::cloneObjectArray`

Wraith0x10 replied:
Ah, I didn't know this method existed. Thanks!

kostamax27 commented on src/event/player/PlayerItemConsumeEvent.php:
```diff
@@ -44,4 +48,32 @@ public function __construct(
+	 *
+	 * @param Item[] $items
+	 */
+	public function setResidue(array $items) : void{
```
`Utils::cloneObjectArray`

Commit:
Use cloneObjectArray (Thanks @kostamax27)

dktapps commented on src/player/Player.php:
```diff
@@ -1727,6 +1727,7 @@ public function consumeHeldItem() : bool{
 			$oldItem = clone $slot;
 
 			$ev = new PlayerItemConsumeEvent($this, $slot);
+			$ev->setResidue([$slot->getResidue()]);
```
An optional constructor parameter would be useful for this

Wraith0x10 replied:
Done

Commit:
Optional parameter in the event

Wraith0x10 commented:
Ready?

dktapps committed:
Update Player.php

dktapps committed:
Update documentation

dktapps committed:
Don't pass air items to the event if the residue is null

when there's no residue, plugins would expect to see an empty array when getting the residue. The old code would have given an array with 1 element containing a null item, which would've been confusing.

dktapps commented:
I've made some tweaks to the docs and also made sure it doesn't pass an air item for regular food. With that I think this is OK

dktapps committed:
Merge branch 'minor-next' into residue-override
